File: programmers/level2/전화번호_목록.java
package level2;

import java.util.HashMap;

public class 전화번호_목록 {
	public static boolean solution(String[] phone_book) {
		boolean answer = true;
		HashMap<String, Object> map = new HashMap<String, Object>();
		for (String str : phone_book) {
			map.put(str, null);
		}
		for (String str : phone_book) {
			for (int i = 0; i < str.length(); i++) {
				if (map.containsKey(str.substring(0, i))) {
					return false;
				}
			}
		}
		return answer;
	}

	public static void main(String[] args) {
		System.out.println(solution(new String[] { "119", "97674223", "1195524421" }));
		System.out.println(solution(new String[] { "123", "456", "789" }));
		System.out.println(solution(new String[] { "12", "123", "1235", "567", "88" }));
		System.out.println(solution(new String[] { "1195524421", "97674223", "119" }));
		System.out.println(solution(new String[] { "113", "44", "4544" }));
		System.out.println(solution(new String[] { "010111", "010" }));
	}
}

File: programmers/level2/전화번호 목록.py
def solution(phone_book):
    # List to set
    dic = {i for i in phone_book}

    # 전화번호 한개씩 꺼냄
    for num in phone_book:
        n = ""
        # 한글자씩 추가해가며 dic에 있는지 확인
        for i in num[:-1]:
            n += i
            # dic에 존재하면 False 반환
            if n in dic:
                return False

    return True


print(solution(["119", "97674223", "1195524421"]))
print(solution(["123", "456", "789"]))
print(solution(["12", "123", "1235", "567", "88"]))
